### PR TITLE
Relax wire_order restrictions in circuit visualization

### PR DESCRIPTION
**Pull request by 1ucian0**
Requested by @ajavadia 

Drawing circuits with `wire_order` has some restrictions, coming from [the original design that](https://github.com/Qiskit/qiskit-terra/issues/8063), I think, can be relaxed a bit. For example if `wire_order` only refers to qubits, then `cregbundle` can be preserver. So this PR relax those restrictions and allows you to write `wire_order=[2, 3, 0, 1]` and `cregbundle=True` in a 4 qubit circuit, for example:

```python
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister

qr = QuantumRegister(4, "q")
cr = ClassicalRegister(4, "c")
cr2 = ClassicalRegister(2, "ca")
circuit = QuantumCircuit(qr, cr, cr2)
circuit.h(0)
circuit.h(3)
circuit.x(1)
circuit.x(3).c_if(cr, 10)

for method in ['text', 'mpl', 'latex']:
    display(circuit.draw(method, wire_order=[2, 3, 0, 1], cregbundle=True))
```

Screenshot 2023-07-10 at 03 50 42

### CODE REVIEW

File: qiskit/opflow/__init__.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 
 Operators and State functions are the building blocks of Quantum Algorithms.
 
-A library for Quantum Algorithms & Applications is more than a collection of
+A library for Quantum Algorithms & Applications (QA&A) is more than a collection of
```
Comment (1ucian0) [inline qiskit/opflow/__init__.py:22]: This is unrelated. I just saw it around and noticed that _QA&A_ is not a very common acronym.

**Review by ajavadia — DISMISSED** [review]
looks great thanks for fixing this

**Review by jakelishman — CHANGES_REQUESTED** [review]
I don't think this is a good idea.  Allowing only a partial `wire_map` I think is very confusing behaviour, and not general like the initial intent of the option - I personally would have expected that giving a partial `wire_map` would _only_ draw those wires, if it was going to have non-error behaviour.  Really though, it just looks like an error to me.

If you _really_ want this behaviuor for `wire_map`, can we at least make the syntax `[2, 3, ...]`?

**Comment by 1ucian0** [comment]
> I personally would have expected that giving a partial `wire_map` would _only_ draw those

To the contrary (and probably because I'm too famiiar with the drawer), only drawing some wires makes little sense to me. (how do you express a `CX(0,1)` without drawing `0` or `1`).

I saw some numpy `Ellipsis` usage in slicing, but it is the first time I see it as an element. Is that a thing? I have no problem to add it.

**Comment by jakelishman** [comment]
We spoke about this in the dev meeting just now.  The notes are:
- Ali's original request was mostly just about allowing _only_ the qubits to be specified with the clbits omitted, so he'd be on board with allowing `wire_map` to have two valid lengths: one with all the qubits and clbits, and one with only all the qubits.  If you prefer that, that there's no need for the `...`.
- Matthew also thought that specifying only a partial list of qubits would have a filtering behaviour.
- If we do want to allow only partial specification of qubits (or partial specification of clbits), there was general support for using the `...` literal as the marker.

Given this, I'd vote to go with option 1: the `wire_order` list has (at most) two valid lengths: "all bits" or "all qubits".  It solves Ali's initial request and it doesn't require extra syntax.  (Personally I was surprised to learn that `wire_order` applies to the clbits at all, and we didn't have separate order fields for qubits and clbits, but that ship has sailed.)

**Comment by jakelishman** [comment]
Oh, and about "how do you draw `cx(0, 1)` if you filter out qubit 1?": that point is mostly why I said "[partial specification] looks like an error", yeah.

**Comment by 1ucian0** [comment]
> Given this, I'd vote to go with option 1: the `wire_order` list has (at most) two valid lengths: "all bits" or "all qubits".

done in that way. Have a look.

File: releasenotes/notes/relax_wire_order_restrictions-ffc0cfeacd7b8d4b.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            ┌───┐ ┌───┐
+       q_3: ┤ H ├─┤ X ├─
+            ├───┤ └─╥─┘
+       q_0: ┤ H ├───╫───
+            ├───┤   ║
+       q_1: ┤ X ├───╫───
+            └───┘┌──╨──┐
+       c: 4/═════╡ 0xa ╞
+                 └─────┘
+      ca: 2/════════════
```
Comment (jakelishman) [inline releasenotes/notes/relax_wire_order_restrictions-ffc0cfeacd7b8d4b.yaml:34]: I don't think this example will work after the new changes any more because of the partial wire order.  We should replace it with one that assigns all qubits.

File: test/python/visualization/test_circuit_text_drawer.py [hunk 3]
```diff
@@ -341,6 +341,37 @@ def test_wire_order(self):
...
+        cr = ClassicalRegister(4, "c")
+        cr2 = ClassicalRegister(2, "ca")
+        circuit = QuantumCircuit(qr, cr, cr2)
+        circuit.h(0)
+        circuit.h(3)
+        circuit.x(1)
+        circuit.x(3).c_if(cr, 10)
+        self.assertEqual(
+            str(_text_circuit_drawer(circuit, wire_order=[2, 1, 3, 0, 4, 5, 6, 7, 8, 9])), expected
+        )
```
Comment (jakelishman) [inline test/python/visualization/test_circuit_text_drawer.py:373]: This test says its `wire_order` is only qubits, but it actually has all bits.

File: test/python/visualization/test_circuit_latex.py [hunk 4]
```diff
@@ -699,6 +699,26 @@ def test_wire_order(self):
...
+        circuit.x(1)
+        circuit.x(3).c_if(cr, 12)
+        circuit_drawer(
+            circuit,
+            cregbundle=False,
+            wire_order=[2, 1, 3, 0, 4, 5, 6, 7, 8, 9],
+            filename=filename,
+            output="latex_source",
+        )
+        self.assertEqualToReference(filename)
```
Comment (jakelishman) [inline test/python/visualization/test_circuit_latex.py:720]: This test says its `wire_order` is only qubits, but it's actually got all of them.  It should probably be a test of what it says, and shouldn't include the clbits.

File: qiskit/opflow/__init__.py (hunk 1 above)
Comment (jakelishman) [reply qiskit/opflow/__init__.py:28]: It'd probably be better to revert this, especially with opflow deprecated.  It's a problem that's existed forever (and capitalising unexpected words in a sentence like this is a not-altogether-unheard-of way of implicitly defining an acronym anyway).

**Review by jakelishman — COMMENTED** [review]
The new logic looks good, thanks, and consistent with what I remember us discussing.  Could you fix the tests so they test what they say they do?

File: qiskit/opflow/__init__.py (hunk 1 above)
Comment (1ucian0) [reply qiskit/opflow/__init__.py:28]: done in 519028529dab09960efba9388b9a2dd77c86aedd

File: test/python/visualization/test_circuit_latex.py (hunk 4 above)
Comment (1ucian0) [reply test/python/visualization/test_circuit_latex.py]: actually, these tests make little sense now. Removing them in 445489c4a8832d9fd73f5a155d21ec3c96f07084 and add those that cover the involved code in this PR.

File: test/python/visualization/test_circuit_text_drawer.py (hunk 3 above)
Comment (1ucian0) [reply test/python/visualization/test_circuit_text_drawer.py]: test removed. See https://github.com/Qiskit/qiskit-terra/pull/9893#discussion_r1262414808

File: releasenotes/notes/relax_wire_order_restrictions-ffc0cfeacd7b8d4b.yaml (hunk 2 above)
Comment (1ucian0) [reply releasenotes/notes/relax_wire_order_restrictions-ffc0cfeacd7b8d4b.yaml:34]: oops... fixed in d2224cdb093515f5eaad1f06c8f5cb72d0c6eff5

**Review by jakelishman — APPROVED** [review]
This looks much better now, thanks.  I think the API we've settled on is the right compromise between permissibility and not being surprising.

**Comment by enavarro51** [comment]
Just a note for all concerned, when we use `circuit_drawer` for tests that are supposed to have "text" output, can we please put `output="text"` in the call for those of us who default to "mpl" or otherwise. Thanks.
